### ISSUE TITLE
rb_sparsity compression loss not changing

### ISSUE THREAD

**Issue by ronandorrepaal** [openvinotoolkit/nncf#1031]
Hello, 

I have been working on the nncf framework with Pytorch to test the sparsification capabilities.

`magniture_sparsity` works well for me, but I have a problem with `rb_sparsity` algorithm. My issue seems to be related to some bug (?) with the compression backprop. For me, the compression loss does not seem to change during training.

I have been working with the CIFAR-100, with a batch size of 256 and my compression loss appears to remain at 36.0000 for a target of 0.7 (sparsity = 30%). 

From what I can tell, the reason seems to be related to the mask saved in `compression_ctrl`. This mask remains unchanged with each training iteration but is used to generate a binary mask, which in turn is used for calculation of the compression loss. Each calculated binary mask generated appears to be completely filled, and therefore have 0% sparsity. In the calculation of the compression loss, the following appears to be happening: 

```
((loss / params - self.target) / self.p).pow(2) =  

(([number_of_ones]/[number_of_elements] – 0.7)/0.05).pow(2) =  

((1 – 0.7)/0.05).pow(2) =  

36.0000 
```
from https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/loss.py#L58

`self.p` is 0.05 and it appears to be set on initialisation.

I have tried running the above with my own minimum working example script and also using the github example classification folder [main.py](https://github.com/openvinotoolkit/nncf/blob/develop/examples/torch/classification/main.py) with [resnet34_cifar_rb_sparsity.json](https://github.com/openvinotoolkit/nncf/blob/develop/examples/torch/classification/configs/sparsity/resnet34_cifar_rb_sparsity.json) on a CIFAR-100 dataset from github and got the same results.  

**A final note**: It appears that during the forward pass, a temporary binary mask is made on the fly and using a random uniform distribution, but I can not see where this mask is saved for calculation of the loss. 

Thank you very much, 

Ronan 

**Appendix**:

Please see an example results below (lines relating to remaining weights and the compression target were added by me): 

 
```
Statistics of the RB-sparsity algorithm:
+----------------------------------------------------------------------+-------+
|                           Statistic's name                           | Value |
+======================================================================+=======+
| A target level of the sparsity for the algorithm for the current     | 0.300 |
| epoch                                                                |       |
+----------------------------------------------------------------------+-------+
| The probability that one weight will be zeroed                       | 0.011 |
+----------------------------------------------------------------------+-------+
: Epoch: [6][0/195] Lr: 0.0001 Time: 1.306 (1.306) Data: 0.574 (0.574) CE_loss: 1.5979 (1.5979) CR_loss: 36.0000 (36.0000) Loss: 37.5979 (37.5979) Acc@1: 54.297 (54.297) Acc@5: 85.547 (85.547)
ctrl remaining weights: 1.0
Compression Target: 0.3
: Epoch: [6][10/195] Lr: 0.0001 Time: 0.699 (0.756) Data: 0.011 (0.062) CE_loss: 1.6053 (1.8154) CR_loss: 36.0000 (36.0000) Loss: 37.6053 (37.8154) Acc@1: 50.391 (50.533) Acc@5: 85.547 (81.037)
ctrl remaining weights: 1.0
Compression Target: 0.3
```
 

Statistics of the sparsified model: 
```
+-----------------------------------------+-------+ 
|            Statistic's name             | Value | 
+=========================================+=======+ 
| Sparsity level of the whole model       | 0     | 
+-----------------------------------------+-------+ 
| Sparsity level of all sparsified layers | 0     | 
+-----------------------------------------+-------+ 
```
 

Statistics by sparsified layers: 
```
+---------------------+------------------+----------------+--------------------+ 
|    Layer's name     |  Weight's shape  | Sparsity level |      Weight's      | 
|                     |                  |                |     percentage     | 
+=====================+==================+================+====================+ 
| ResNet/NNCFConv2d[c | [64, 3, 3, 3]    | 0              | 0.008              | 
| onv1]/conv2d_0      |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [64, 64, 3, 3]   | 0              | 0.173              | 
| ayer1]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [64, 64, 3, 3]   | 0              | 0.173              | 
| ayer1]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [64, 64, 3, 3]   | 0              | 0.173              | 
| ayer1]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [64, 64, 3, 3]   | 0              | 0.173              | 
| ayer1]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [64, 64, 3, 3]   | 0              | 0.173              | 
| ayer1]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [64, 64, 3, 3]   | 0              | 0.173              | 
| ayer1]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 64, 3, 3]  | 0              | 0.346              | 
| ayer2]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 64, 1, 1]  | 0              | 0.038              | 
| ayer2]/BasicBlock[0 |                  |                |                    | 
| ]/Sequential[shortc |                  |                |                    | 
| ut]/NNCFConv2d[0]/c |                  |                |                    | 
| onv2d_0             |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[3 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [128, 128, 3, 3] | 0              | 0.692              | 
| ayer2]/BasicBlock[3 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 128, 3, 3] | 0              | 1.384              | 
| ayer3]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 128, 1, 1] | 0              | 0.154              | 
| ayer3]/BasicBlock[0 |                  |                |                    | 
| ]/Sequential[shortc |                  |                |                    | 
| ut]/NNCFConv2d[0]/c |                  |                |                    | 
| onv2d_0             |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[3 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[3 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[4 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[4 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[5 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [256, 256, 3, 3] | 0              | 2.768              | 
| ayer3]/BasicBlock[5 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 256, 3, 3] | 0              | 5.535              | 
| ayer4]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 512, 3, 3] | 0              | 11.071             | 
| ayer4]/BasicBlock[0 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 256, 1, 1] | 0              | 0.615              | 
| ayer4]/BasicBlock[0 |                  |                |                    | 
| ]/Sequential[shortc |                  |                |                    | 
| ut]/NNCFConv2d[0]/c |                  |                |                    | 
| onv2d_0             |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 512, 3, 3] | 0              | 11.071             | 
| ayer4]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 512, 3, 3] | 0              | 11.071             | 
| ayer4]/BasicBlock[1 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 512, 3, 3] | 0              | 11.071             | 
| ayer4]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv1] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/Sequential[l | [512, 512, 3, 3] | 0              | 11.071             | 
| ayer4]/BasicBlock[2 |                  |                |                    | 
| ]/NNCFConv2d[conv2] |                  |                |                    | 
| /conv2d_0           |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
| ResNet/NNCFLinear[l | [100, 512]       | 0              | 0.240              | 
| inear]/linear_0     |                  |                |                    | 
+---------------------+------------------+----------------+--------------------+ 
 ```

**Comment by ljaljushkin** [comment]
Hello, @ronandorrepaal! 

I reproduced your issue and was able to overcome it by setting lower initial sparsity rate.
The issue is not related to some in backward compatibility, because rb_sparsity works well for [MobileNet-v2](https://github.com/openvinotoolkit/nncf/blob/develop/tests/torch/data/configs/weekly/classification/cifar100/mobilenet_v2_rb_sparsity_int8.json) in our nightly validation.


To prove that there's no issue with backprop I intentionally made an initial mask with all ones to prune all weighs. And the backprop quickly reduce mask values. sparsity level and compression loss decreased very quickly.


Please try the following multistep sparsity schedule.

```json
"compression": {
    "algorithm": "rb_sparsity",
    "params": {
       "schedule": "multistep",
       "multistep_steps": [
           16,
           24,
           32,
           40
       ],
       "multistep_sparsity_levels": [
           0.1,
           0.2,
           0.3,
           0.4,
           0.52
       ],
    }
}
```

Regarding your note:

> A final note: It appears that during the forward pass, a temporary binary mask is made on the fly and using a random uniform distribution, but I can not see where this mask is saved for calculation of the loss.

Mask with random noise is going to the output of model and to the cross entropy loss. 
Mask without random noise is explicitly taken for compression loss calculation: https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/sparsity/rb/loss.py#L51
